upup/pkg/fi/cloudup/docker.go

```diff
@@ -0,0 +1,239 @@
+	if c.Spec.Docker == nil || fi.StringValue(c.Spec.Docker.Version) == "" {
+		return nil, nil, fmt.Errorf("unable to find Docker version")
+	}
+
+	version := fi.StringValue(c.Spec.Docker.Version)
+
+	assetUrl, assetHash, err := findDockerVersionUrlHash(arch, version)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return findAssetsUrlHash(assetBuilder, assetUrl, assetHash)
+}
+
+func findDockerVersionUrlHash(arch architectures.Architecture, version string) (u string, h string, e error) {
+	dockerAssetUrl, err := findDockerVersionUrl(arch, version)
+	if err != nil {
+		return "", "", err
+	}
+	dockerAssetHash, err := findDockerVersionHash(arch, version)
+	if err != nil {
+		return "", "", err
+	}
+
+	return dockerAssetUrl, dockerAssetHash, nil
+}
+
+func findDockerVersionUrl(arch architectures.Architecture, version string) (string, error) {
+	sv, err := semver.ParseTolerant(version)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse version string: %q", version)
+	}
+	if sv.LT(semver.MustParse("17.3.0")) {
+		return "", fmt.Errorf("unsupported legacy Docker version: %q", version)
+	}
+
+	var u string
+	switch arch {
+	case architectures.ArchitectureAmd64:
+		if sv.GTE(semver.MustParse("18.9.0")) {
+			u = fmt.Sprintf(dockerVersionUrlAmd64, version)
+		} else {
+			u = fmt.Sprintf(dockerLegacyUrlAmd64, version)
+		}
+	case architectures.ArchitectureArm64:
+		if sv.GTE(semver.MustParse("18.9.0")) {
+			u = fmt.Sprintf(dockerVersionUrlArm64, version)
+		} else if sv.GTE(semver.MustParse("17.9.0")) {
+			u = fmt.Sprintf(dockerLegacyUrlArm64, version)
+		} else {
+			u = fmt.Sprintf(dockerLegacyUrlArm64, dockerFallbackVersion)
+		}
+	default:
+		return "", fmt.Errorf("unknown arch: %q", arch)
+	}
+
+	return u, nil
+}
+
+func findDockerVersionHash(arch architectures.Architecture, version string) (string, error) {
+	sv, err := semver.ParseTolerant(version)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse version string: %q", version)
+	}
+	if sv.LT(semver.MustParse("17.3.0")) {
+		return "", fmt.Errorf("unsupported legacy Docker version: %q", version)
+	}
+
+	var h string
+	switch arch {
+	case architectures.ArchitectureAmd64:
+		h = findAllDockerHashesAmd64()[version]
+	case architectures.ArchitectureArm64:
+		if sv.GTE(semver.MustParse("17.9.0")) {
+			h = findAllDockerHashesArm64()[version]
+		} else {
+			h = findAllDockerHashesArm64()[dockerFallbackVersion]
+		}
+	default:
+		return "", fmt.Errorf("unknown arch: %q", arch)
+	}
+
+	if h == "" {
+		return "", fmt.Errorf("unknown hash for Docker version: %s - %s", arch, version)
+	}
+
+	return h, nil
+}
+
+func findAllDockerHashesAmd64() map[string]string {
+	hashes := map[string]string{
+		"17.03.0":  "aac08524db82d3fdc8fc092f495e1174f5e1dd774b95a6d081544997d34b4855",
+		"17.03.1":  "3e070e7b34e99cf631f44d0ff5cf9a127c0b8af5c53dfc3e1fce4f9615fbf603",
+		"17.03.2":  "183b31b001e7480f3c691080486401aa519101a5cfe6e05ad01b9f5521c4112d",
```

Comment:
So moving all distributions to the "static" variant. Is the removal of the docker-ce-selinux package on RHEL-7 and derived distributions not a problem?

(This is a change for 18.06.3 and earlier, with the exception of 17.03.2 on Ubuntu 18.04.)

Reply:
@justinsb also noticed this, so will try to explain below.

Kops support for SELinux was added in 1.19 (mostly experimental from my point of view) and could not be enabled in docker config in the past. Contents of the packages show that it contains only the SELinux profiles:
```bash
> rpm -qlp ./docker-ce-selinux-17.03.2.ce-1.el7.centos.noarch.rpm 
/usr/share/doc/docker-ce-selinux-17.03.2.ce
/usr/share/doc/docker-ce-selinux-17.03.2.ce/LICENSE
/usr/share/selinux/devel/include/services/docker.if
/usr/share/selinux/packages/docker.pp.bz2
```

The `docker-ce-selinux` package is distributed only for 17.03.x and later versions switch to the `container-selinux` dependency, distributed with RHEL/CentOS.

Considering that SELinux could not be used in the past and we already install by default a replacement package, I don't think this is an issue.